### ISSUE TITLE
Lack of documentation regarding unit testing.

### ISSUE THREAD

**Issue by jforseth210** [flask-extensions/flask_simplelogin#20]
After putting it off for way too long, I've finally decided to teach myself unit testing with the unittest module and write some tests for my Flask project. I've run into a bit of a roadblock with the ```@login_required``` decorator though. I can't figure out how to test any of my views that require a login.
For all of my other views, I'm able to use ```app.test_client(self).get() or .post()``` to make a request and run assertions on the response. However, I can't seem to get that to work with my login form since I don't have a valid csrf token. I saw that you had a similar test in [test_app.py](https://github.com/flask-extensions/flask_simplelogin/blob/master/tests/test_app.py) but I had some trouble modifying the session in my code, and I saw your comment:
>\#token is still invalid :(

leading me to believe that this was a dead end. 

Next, I searched this repo for references to csrf tokens, and didn't find anything. I realized that this was probably handled by WTForms, so I searched their docs for anything to do with testing, but the only thing I found was talking about recaptcha which didn't help me at all. 

I came back to this repo and noticed that there's the ```basic``` parameter for ```@login_required()``` so I tried to figure out a way to only enable basic logins when I was running a test. Modifying all of my ```@login_required``` decorators to accept a boolean seemed hacky, and I ran into circular import problems between my main file and my blueprints.

At this point, I'm using the ```LiveServerTestCase``` from flask_testing with Selenium. It works, but running a whole browser is slow and seems hacky. 

The relevant code from my test file:
```python
import unittest
import time
from flask_testing import LiveServerTestCase
from selenium import webdriver
from selenium.webdriver.common.keys import Keys
from webtool import app

# Doesn't work
class FlaskTestCase(unittest.TestCase):
    def test_login(self):
        tester = app.test_client(self)
        response = tester.post(
            '/login/', data=dict(username='testing', password=PROJECT_PASSWORD, next='/'))
        # csrf token missing

# Works but inefficent.
class LiveServer(LiveServerTestCase):
    def create_app(self):
        app.config['TESTING'] = True
        app.config['LIVESERVER_PORT'] = 0
        return app

    def test_correct_login(self):
        driver = webdriver.Firefox()
        driver.get(self.get_server_url()+"/login")
        username_elem = driver.find_element_by_id('username')
        username_elem.send_keys(username)
        password_elem = driver.find_element_by_id('password')
        password_elem.send_keys(password)
        password_elem.send_keys(Keys.RETURN)
        time.sleep(2)  # Give it time to load
        self.assertTrue('Successful' in driver.page_source)
        driver.close()
```
tl;dr: I can't figure out how to unit test views protected by ```@login_required()```. I was able to get it working with Selenium, but it's too slow.

**Comment by jforseth210** [comment]
Nevermind, I figured it out. For anyone having the same problem, check out this Stackoverflow answer:
[https://stackoverflow.com/questions/21506326/testing-flask-login-and-authentication](https://stackoverflow.com/questions/21506326/testing-flask-login-and-authentication)
The Realpython article he links to is also really helpful:
[https://realpython.com/blog/python/python-web-applications-with-flask-part-iii/](https://realpython.com/blog/python/python-web-applications-with-flask-part-iii/)

**Comment by cuducos** [comment]
Hey, @jforseth210  — good to know you’ve found a way, and even better to know your studying tests!

If you allow me, I’ll share my two cents on the CSRF and tests.

This is not the repo directly related to your issue. The [part that implements the CSRF token in the login form is the Flask-WTF](https://flask-wtf.readthedocs.io/en/stable/csrf.html). And as they say:

> You can disable CSRF protection in all views by default, by setting `WTF_CSRF_CHECK_DEFAULT` to False

Thus, you can just disable the CSRF for the sake of testing if this layer of security doesn’t add much to your test suite. Once you create the Flask application, you can change its config, something like:



```python
class FlaskTestCase(unittest.TestCase):
    
    def setUp(self):
        app.config["TESTING"] = True
        app.config["WTF_CSRF_CHECK_DEFAULT"] = False
        self.client = app.test_client()

    def test_login(self):
        data = {"username": "testing", "password": PROJECT_PASSWORD, "next": "/"}       
        response = self.client.post("/login/", data=data)
        # TODO (shouldn’t comlain about CSRF as WTF_CSRF_CHECK_DEFAULT is False
```

**Comment by oegedijk** [comment]
New to Flask testing, so it's still not clear how to test whether the page actually renders correctly after login.

```
data = {"username": "user", "password": "password", "next":"/"}       
with self.client as client:
  response1 = client.post("/login/", data=data, follow_redirects=True)
  response2 = client.get("/", follow_redirects=True)
```

Both come `response1.data` and `response2.data` just give back the form html:

```
b'<!DOCTYPE html>\n<html lang="en">\n<head>\n  <meta charset="UTF-8">\n  <meta name="viewport" content="width=device-width, initial-scale=1.0">\n  <meta http-equiv="X-UA-Compatible" content="ie=edge">\n  <title>Flask Login</title>\n  <link rel="stylesheet" href="https://maxcdn.bootstrapcdn.com/bootstrap/4.0.0-beta/css/bootstrap.min.css" integrity="sha384-/Y6pD6FV/Vv2HJnA6t+vslU6fwYXjCFtcEpHbNJ0lyAFsXTsjBbfaDjzALeQsN6M" crossorigin="anonymous">\n</head>\n<body>\n    \n    \n      <ul class="flashes alert">\n      \n        <li class="flashes alert alert-warning">You need to login first</li>\n      \n      </ul>\n    \n    \n\n    \n\n    <form action="/login/" method="post">\n      <div class="form-group">\n      <input id="csrf_token" name="csrf_token" type="hidden" value="IjQ4ZjFjZDY3ZDAzMjBkOTMxMzM5NjEwNjkwNjZmMDc4YjIzYzUwNGQi.X_LtTw.YHNFJFOebhfODyTgfu9rHS5YccE">\n      <label for="username">name</label><div class="form-control"><input id="username" name="username" required type="text" value=""></div><br>\n      <label for="password">password</label><div class="form-control"> <input id="password" name="password" required type="password" value=""></div><br>\n      <input type="hidden" name="next" value="/_dashboard1">\n      </form>\n     <input type="submit" value="Send">\n    </form>\n</body>\n</html>'
```

How do I test for the page behind the `login_required` decorator?

**Comment by cuducos** [comment]
> How do I test for the page behind the `login_required` decorator?

A view decorated with `login_required` should redirect to `/login/`. So I wouldn't use the `follow_redirects=True` and just check (a) if the status code is 302 and (b) the location (in the HTTP response) is `/login/`. Does that make sense?

To be honest, if you're studying tests in web development, this is mostly a matter of HTTP than of Flask itself — saying that so you know what to cover ; )

**Comment by oegedijk** [comment]
Ah, so probably best would be:

1) test a version without logins to see if all the page correctly load, etc.
2) test with logins to see if the pages redirect with code 302

I'm a data scientist building a library to quickly build explainable AI dashboards ([https://github.com/oegedijk/explainerdashboard](https://github.com/oegedijk/explainerdashboard)), so web dev is not really my strong suit, but needed for this project. Your library really helped to make adding login functionality easier, and also to allow me to wrap the whole dashboard into a single class: [ExplainerHub](https://explainerdashboard.readthedocs.io/en/latest/hub.html).

So thanks for building it!

**Comment by cuducos** [comment]
> Ah, so probably best would be:
>
> 1. test a version without logins to see if all the page correctly load, etc.
> 2. test with logins to see if the pages redirect with code 302

Excellent. Just a reminder that when tests expect 302 (redirect) it's a best practice to _also_ assert the location to make sure it's not a false-positive (redirecting but to an unexpected location).

**Comment by jforseth210** [comment]
You can try this function I wrote for logins, but bear in mind that it was written by a self-taught teenager and that the quality reflects that:
[https://github.com/jforseth210/jforseth.tech/blob/master/tests/test_accounts.py](https://github.com/jforseth210/jforseth.tech/blob/master/tests/test_accounts.py)
The login method basically does exactly what logging in with a browser would do: It gets the login page, gets the token, and sends a post request with the token, username, and password. 
Let me know if I can help!
